src/audio/copier/copier.c

```diff
@@ -748,6 +755,82 @@ static int set_attenuation(struct comp_dev *dev, uint32_t data_offset, const cha
 	return 0;
 }
 
+static int set_chmap(struct comp_dev *dev, const void *data, size_t data_size)
+{
+	const struct ipc4_copier_config_channel_map *chmap_cfg = data;
+	struct processing_module *mod = comp_mod(dev);
+	struct copier_data *cd = module_get_private_data(mod);
+	enum ipc4_direction_type dir;
+	struct ipc4_audio_format in_fmt = cd->config.base.audio_fmt;
+	struct ipc4_audio_format out_fmt = cd->config.out_fmt;
+	pcm_converter_func process;
+	pcm_converter_func converters[IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT];
+	int i;
+	uint32_t irq_flags;
+
+	if (data_size < sizeof(*chmap_cfg)) {
+		comp_err(dev, "Wrong payload size: %d", data_size);
+		return -EINVAL;
+	}
+
+	if (cd->endpoint_num == 0 || dev->ipc_config.type != SOF_COMP_DAI) {
+		comp_err(dev, "Only DAI gateway supports changing chmap");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "New chmap requested: %x", chmap_cfg->channel_map);
+
+	if (!cd->dd[0]->dma_buffer) {
+		/* DMA buffer not yet created. Remember the chmap, it will be used
+		 * later in .params() handler.
+		 *
+		 * The assignment should be atomic as LL thread can preempt this IPC thread.
+		 */
+		cd->dd[0]->chmap = chmap_cfg->channel_map;
+		return 0;
+	}
+
+	copier_dai_adjust_params(cd, &in_fmt, &out_fmt);
+
+	dir = (cd->direction == SOF_IPC_STREAM_PLAYBACK) ?
+		ipc4_playback : ipc4_capture;
+
+	process = get_converter_func(&in_fmt, &out_fmt, cd->gtw_type, dir, chmap_cfg->channel_map);
+
+	if (!process) {
+		comp_err(dev, "No gtw converter func found!");
+		return -EINVAL;
+	}
+
+	/* Channel map is same for all sinks. However, as sinks allowed to have different
+	 * sample formats, get new convert/remap function for each sink.
+	 */
+	for (i = 0; i < IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT; i++) {
+		if (cd->converter[i]) {
+			converters[i] = get_converter_func(&in_fmt, &cd->out_fmt[i],
+							   ipc4_gtw_none, ipc4_bidirection,
+							   chmap_cfg->channel_map);
+			/* Do not report an error if converter not found as sinks could be
+			 * bound/unbound on a fly and out_fmt[i] may contain obsolete data.
+			 */
+		} else {
+			converters[i] = NULL;
+		}
+	}
+
+	/* Atomically update chmap, process and converters */
+	irq_local_disable(irq_flags);
```

Comment:
what is this protecting against? No racing against other cores is possible?

Reply:
set_chmap() is called from IPC thread. It could be called (and is intended to be called) when pipeline is running and so it could be preempted by LL thread any time. Copier has up to 4 outputs.
Consider the case: let's say we have copier with 2 sinks; before set_chmap() dummy chmap was used, i.e., no remapping and audio_stream_copy() is used as conversion function for all copier sinks; then set_chmap() is called from IPC thread with channel map requesting to swap channels, remap_c16() func is assigned as conversion func to first sink, but before it is also assigned to second sink, LL thread preempts IPC thread and runs copier process func with remap_c16() assigned for sink 1 and audio_stream_copy() still remained for sink 2. Now copier produces two different streams one with swapped channels and another without swapping.
Actually, seems assigning channel map could be removed from that guarded block, but assigning process and conversion funcs should be atomic.

Reply:
@serhiy-katsyuba-intel sorry, I meant - is just locking interrupts enough or do we need a spinlock? Are you relying on the fact that if that pipeline is running on a secondary core, the IPC thread will also run on the same core? What about cross-core DP streams?